### DISCUSSION TITLE
Doorbell - minimum and maximum ringing time rule

### DISCUSSION THREAD

**Discussion by Seba200086** [arendst/Tasmota#13094] [Q&A] [answered]
Hi, 

I would like to convert my dumb doorbell into a smart device. I know there are a lot of guides on the internet about this, however, I don't know how to cover all of my assumptions....

Hardware:
Bell (230V) connected to a push-button (NO)
Shelly 1 with Tasmota 9.5.0

Assumptions: 
1. Switch configured as a button - bell rings as long as button is pressed - SwitchMode 7 (clear)
2. Possibility to switch off the bell at certain times (e.g. 10 PM to 7 AM) and leave notifications (MQTT) - SwitchMode 15 (clear)
3. Possibility to use the bell as an alarm - I will send a command via MQTT (clear)

4. The bell rings for a set period of time:
a. button pressed <2s - bell rings for 2s (e.g. someone pressed the button for 0.5s - bell rings for 2s anyway)
b. button pressed from 2 to 4s - chime rings as long as button is pressed (e.g. 3s)
c. button pressed >4s - the bell rings only 4s
d. every time after the key was pressed (ringing) there is a pause (the bell cannot be rung) for 4 seconds

I think I can handle assumptions 1-3 but I have no idea how to do point 4.
The whole thing will work with Home Assistant, but I think this point should be implemented as appropriate rule(s) directly in Tasmota (to avoid problems with HA <-> Tasmota communication).

Could someone help me to prepare a suitable rule?

**Comment by barbudor** [comment, marked as answer]
HI
Aren't you over-engineering this ? 😂

This is my guess. I didn't tested it for real but it looks like it does what you need
```
switchmode1 1   (or maybe switchmode1 2 - try for yourself)
Rule1
  on switch1#state=1 do backlog var1 1; var2 1; power on; ruletimer1 2; ruletimer2 4 endon
  on switch1#state=0 do backlog var1 0 ; if (%var2%=0) power off; ruletimer0 0 endon
  on rules#timer=1 do backlog var2 0; if(%var1%=0) power off; ruletimer0 0 endon
  on rules#timer=2 do power off endon
```
Note that you will need to self compile a version with support for IF, or use `tasmota-fullrules` from [here](https://github.com/tasmota/install/tree/main/firmware/unofficial)

Some explanations:
As you can't test the state of a switch or of a timer, I'm using:
- `var1` to track the state of the switch
- `var2` to track the state of the RuleTimer1
- Note the syntax `Ruletimer0 0` which stops and clear all running timers in a single call

I'll let you read that and test it.

**Comment by Seba200086** [comment]
First of all, thank you very much for your reply and extensive explanation!

I don't have any experience of setting up rules for Tasmota but I've looked through the code (which actually seems pretty straightforward once you have some experience) and it seems like it should work.

Unfortunately, I bought Shelly 1 with Tasmota preloaded.
Today I ordered a programmer (TTL), when it arrives I will upload `tasmota-fullrules` as you suggested.

I understand, that in explanations instead of "Note the syntax `Rule0 0`" you meant "Note the syntax `ruletimer0 0`".

What I am still missing to be completely happy is the blocking of the possibility to ring for 4s after the last ring off (after the last `power off` command).

Do you see a simple solution?

**Reply by sfromis** [reply]
You should  be able to upgrade to fullrules without wired flashing. If you download it from [here](https://github.com/tasmota/install/blob/main/firmware/unofficial/tasmota-fullrules.bin), and also download the -minimal binary, you should be able to do the two-step upgrade using file upload. It would also be possible to change the rules to not need IF, but it would get a bit more complex.

**Reply by Seba200086** [reply]
Thank you for the hint. Sometimes the simplest things are the hardest to see.

Just to make sure:
0. I have Tasmota 9.5.0 (standard) uploaded
1. First upload [minimal](https://github.com/tasmota/install/blob/main/firmware/release/tasmota-minimal.bin)
2. Next [fullrules](https://github.com/tasmota/install/blob/main/firmware/unofficial/tasmota-fullrules.bin)

Not the other way around?
On Tasmota 9.5.0 cannot `fullrules `be uploaded directly?

**Reply by sfromis** [reply]
You have the sequence right, minimal is a smaller binary specifically for upgrading when the larger one would not fit. However, you can actually upload fullrules "directly" OTA, **if** you do it from a http server, not from https (which is not supported by the upgrade process). When doing OTA from a http server, Tasmota will automagically use the minimal extra step when needed.

**Comment by Seba200086** [comment]
OK, I think I have understood a bit of the syntax of the rules thanks to your help.

I added support for cool down after the last power off and changed the syntax a bit - threw if's into new lines - according to the [rules documentation](https://tasmota.github.io/docs/Rules/#conditional-rules).

```
Rule1
  ON system#boot DO backlog var1 0; var3 0; var5 0 ENDON
  ON switch1#state=1 DO if (%var3%!=1) var5 1; var1 1; power on; ruletimer1 2; ruletimer2 4 endif ENDON
  ON switch1#state=0 DO var5 0 ENDON
  ON switch1#state=0 DO if (%var1%=0) power off; ruletimer3 4; var3 1; ruletimer1 0; ruletimer2 0 endif ENDON
  ON rules#timer=1 DO var1 0 ENDON
  ON rules#timer=1 DO if(%var5%=0) power off; ruletimer3 4; var3 1; ruletimer2 0 endif ENDON
  ON rules#timer=2 DO backlog power off; ruletimer3 4; var3 1 ENDON
  ON rules#timer=3 DO var3 0 ENDON
```

Where:
`var5` - track the state of the switch (5witch)
`var1` - track the state of the RuleTimer1
`var3` - track the state of the RuleTimer3

`ruletimer1` - timer for minimum ring time (2s)
`ruletimer2` - timer for maximum ring time (4s)
`ruletimer3` - cool down timer (between rings) (4s)

I also think I might remove the % characters from the variables.

I will try to test the operation later and let you know.

**Reply by barbudor** [reply]
You may want to invent new ways or syntax, but only things that are correct will work at the end
The % % are mandatory per rule syntax
Duplicating the same trigger is just losing efficiency and bring no values

**Reply by Seba200086** [reply]
Hi Barbudor,
As I wrote, I changed the syntax according to Tasmota documentation.
Perhaps I don't fully understand it (this eventuality is highly probable :) ).

Especially those parts:

```
NOTE
Same trigger may be used in more than one rule. This may be required for some cases of using IF/ELSE since an IF statement cannot be used within a Backlog.
```
and
```
But not like this: 
Power1 off; LedPower on; IF (Mem1==0) Var1 Var1+1; Mem1 1 ENDIF; Delay 10; Power1 on

You should split it in two lines like: 
ON Power2#state=1 DO Power1 off; LedPower on; ENDON 
ON Power2#state=1 DO IF (Mem1==0) Var1 Var1+1; Mem1 1 ENDIF; Delay 10; Power1 on ENDON
```

As you can see there is a command after if moved to a separate line.

Just below is also an example of usage, where there is no % around variables.
```
Rule1
 ON system#boot DO var1 0 ENDON
 ON energy#power>100 DO if (var1!=1) ruletimer1 0;var1 1 endif ENDON
 ON tele-energy#power<50 DO if (var1==1) var1 2;ruletimer1 600 endif ENDON
 ON rules#timer=1 DO Backlog var1 0;power off ENDON 
```

But these are details of course. Thanks again for your help!

**Comment by Seba200086** [comment]
Thanks to your help and instructions I changed Tasmota to fullrules version and ran this rule.

Theoretically it works.

Practically, I get strange behaviours - due to delays in the operation of the rules, which I did not expect at all.

For testing I connected a light bulb instead of a bell. 

During tests for the version with cool down after switching off, from pressing the button to lighting the bulb almost 2s(!) pass (in the console you can see consecutive steps - first if, then var 5=1, var1=1, at the end only power on...)
Of course, I can change the order (power on first, then the variables) - it's a little bit better (but still ~1s lag), but then, when the button is pressed shortly (quick on/off) before the variables are set to 1, another rule starts (switch1#state=0 - set var5=0), but after a while var5=1 is present (the commands from the first line are executed)...

Are delays in rules execution normal? Is it possible to do something about it?
